cli/src/commands/update/dataset.rs

```diff
@@ -0,0 +1,70 @@
+use anyhow::{Context, Result};
+use log::info;
+use reinfer_client::{Client, DatasetIdentifier, SourceId, SourceIdentifier, UpdateDataset};
+use structopt::StructOpt;
+
+/// Update a dataset.
+#[derive(Debug, StructOpt)]
+pub struct UpdateDatasetArgs {
+    #[structopt(name = "dataset")]
+    /// Name or id of the dataset to delete
+    dataset: DatasetIdentifier,
+
+    #[structopt(long = "title")]
+    /// Set the title of the dataset
+    title: Option<String>,
+
+    #[structopt(long = "description")]
+    /// Set the description of the dataset
+    description: Option<String>,
+
+    #[structopt(short = "s", long = "source")]
+    /// Names or ids of the sources in the dataset
+    sources: Option<Vec<SourceIdentifier>>,
+}
+
+pub fn update(client: &Client, args: &UpdateDatasetArgs) -> Result<()> {
+    let UpdateDatasetArgs {
+        ref dataset,
+        ref title,
+        ref description,
+        ref sources,
+    } = *args;
+
+    let source_ids = sources
+        .as_ref()
+        .map::<Result<Vec<SourceId>>, _>(|sources| {
+            sources
+                .iter()
+                .map(|source| Ok(client.get_source(source.clone())?.id))
+                .collect()
+        })
+        .transpose()
+        .context("Operation to get sources failed")?;
+
+    let dataset_full_name = match dataset.to_owned() {
+        DatasetIdentifier::FullName(name) => name,
```

Comment:
is `from_str` constructor called automatically if the argument is a string?

Reply:
I'm not sure what you mean here? `DatasetIdentifier::FullName(name)` is a pattern that allows us to take the inner value `name` out of the enum variant `DatasetIdentifier::FullName`. There's no construction here.

`from_str` is called automatically by `StructOpt` when parsing the cli arguments into the `args`, at the beginning of the program.